src/machine/asm/execute_x64.S

```diff
@@ -1251,6 +1254,260 @@ ckb_vm_x64_execute:
   WRITE_RD(RS1)
   NEXT_INST
 .p2align 3
+.CKB_VM_ASM_LABEL_OP_LR_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  movq RS1, LOAD_RESERVATION_ADDRESS
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_SC_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_WRITE(RS1, TEMP1d, 4)
+  movq LOAD_RESERVATION_ADDRESS, TEMP1
+  cmp RS1, TEMP1
+  jne .sc_w_branch
+  movq RS2r, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  movq $UINT64_MAX, LOAD_RESERVATION_ADDRESS
+  WRITE_RD_VALUE($0, TEMP1)
+  NEXT_INST
+.sc_w_branch:
+  WRITE_RD_VALUE($1, TEMP1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOSWAP_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_WRITE(RS1, TEMP1d, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOADD_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_WRITE(RS1, TEMP1d, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  add TEMP1, RS2r
```

Comment:
I noticed a line in RISC-V's spec:

`For RV64, 32-bit AMOs always sign-extend the value placed in rd, and ignore the upper 32 bits of the original value of rs2`

Does this mean we need to clear the upper 32 bits of `rs2` here for all `AMO*.W` operations?

Reply:
After `add TEMP1, RS2r`, we write the result to memory using the following statement:

```
mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
```

It has no affects the final result whether or not the upper 32 bits of rs2 are cleared.

Reply:
2 things:

* The spec requires `rd` to be sign-extended, which means `mov RS2rd` might not be correct here?
* For `add` you are correct, this might end up in the same result, but what about `MIN`/`MAX`? Ignoring upper 32 bits or sign-extending can lead to different result, correct?

Reply:
- `mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)` moves RS2rd to memory, not `rd`, and rd = sign_ext(memory[rs1])
- Good question, I need to take a closer look.